File: 739.cpp
#include <vector>
using namespace std;

class Solution {
private:
  struct Datum {
    int val;
    int index;
  };

public:
  vector<int> dailyTemperatures(vector<int> &temperatures) {
    vector<Datum> s;
    vector<int> ans(temperatures.size());
    int i = 0;
    for (auto temp : temperatures) {
      while (!s.empty() && s.back().val < temp) {
        ans[s.back().index] = i - s.back().index;
        s.pop_back();
      }
      s.push_back({temp, i});
      ++i;
    }
    return ans;
  }
};

File: 739.py
class Solution:
    def dailyTemperatures(self, temperatures: list[int]) -> list[int]:
        # stack stores:
        # tuple: 0th = value of temperatures[i], 1st: i
        s = []
        ans = [0] * len(temperatures)
        for idx, temp in enumerate(temperatures):
            while s and temp > s[-1][0]:
                ans[s[-1][1]] = idx - s[-1][1]
                s.pop()
            else:
                s.append((temp, idx))
        return ans
